File: JavaScript/212-Word Search II.js
/**
 * @param {character[][]} board
 * @param {string[]} words
 * @return {string[]}
 */
var treeNode = function(){
    this.children = new Map();
    this.isWord = false;
}

var findWords = function(board, words) {
    let root = new treeNode();
    
    for (let word of words){
        let node = root;
        for (let char of word){
            if (!node.children.has(char)){
                node.children.set(char, new treeNode());
            }
            node = node.children.get(char);
        }
        node.isWord = word;
    }
    
    // let printTrie = (node) =>{
    //     if (node.isWord){
    //         console.log(node.isWord);
    //     }
    //     for (let nxt of node.children.keys()){
    //         printTrie(node.children.get(nxt));
    //     }
    // }
    // printTrie(root);
    
    let res = new Set();
    let row = board.length, col = board[0].length;
    let dfs = (r, c, node) =>{
        if (node.isWord){
            res.add(node.isWord);
        }
        
        for (let [dr, dc] of [[1,0], [0,1], [-1,0], [0,-1]]){
            let newR = r+dr, newC = c+dc;
            if (0<=newR && newR<row && 0<=newC && newC<col && node.children.has(board[newR][newC])){
                let tmp = board[newR][newC];
                board[newR][newC] = '#';
                dfs(newR, newC, node.children.get(tmp));
                board[newR][newC] = tmp;
            }
        }
    }
    
    for (let r=0; r<row; r++){
        let currNode = root;
        for (let c=0; c<col; c++){
            if (currNode.children.has(board[r][c])){
                let tmp = board[r][c];
                board[r][c] = '#';
                dfs(r, c, currNode.children.get(tmp));
                board[r][c] = tmp;
            }
        }
    }
    return [...res];
};

File: C++/212-Word Search II.cpp
struct TrieNode {
    TrieNode* children[26];
    string s = "";
    explicit TrieNode(){
        std::fill(children, children +26 , nullptr);
    }
    virtual ~TrieNode(){ // this makes code even faster
        for(TrieNode* e : children) delete e;
    }
};

class Solution {
public:
    int row, col;
    vector<string> res;
    vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        this->row = board.size();
        this->col = board[0].size();
        TrieNode* root = new TrieNode();
        for (string word:words) {
            TrieNode* node = root;
            for (char w:word) {
                if (node->children[w-'a'] == nullptr) {
                    node->children[w-'a'] = new TrieNode();
                }
                node = node->children[w-'a'];
            }
            node->s = word;
        }

        for (int r=0; r<row; r++) {
            for (int c=0; c<col; c++) {
                if (root->children[board[r][c]-'a'] == nullptr) continue;
                vector<vector<bool>> visited(row, vector<bool>(col, false));
                visited[r][c] = true;
                dfs(root->children[board[r][c]-'a'], r, c, board, visited);
            }
        }

        return res;
    }
    void dfs(TrieNode* node, int r, int c, vector<vector<char>>& board, vector<vector<bool>>& visited) {
        if (!((node->s).empty())) {
            res.push_back(node->s);
            node->s = "";     // note the trick here
        }
        for (auto &[dr, dc]:directions) {
            int nr = r+dr, nc = c+dc;
            if (0<=nr && nr<row && 0<=nc && nc<col && !visited[nr][nc] && node->children[board[nr][nc]-'a'] != nullptr) {
                visited[nr][nc] = true;
                dfs(node->children[board[nr][nc]-'a'], nr, nc, board, visited);
                visited[nr][nc] = false;
            }
        }
    }
};



// Better, without using vector<vector<bool>> visited
struct TrieNode {
    TrieNode* children[26];
    string s = "";
    explicit TrieNode(){
        std::fill(children, children +26 , nullptr);
    }
    virtual ~TrieNode(){ // this makes code even faster
        for(TrieNode* e : children) delete e;
    }
};

class Solution {
public:
    int row, col;
    vector<string> res;
    vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        this->row = board.size();
        this->col = board[0].size();
        TrieNode* root = new TrieNode();
        for (string word:words) {
            TrieNode* node = root;
            for (char w:word) {
                if (node->children[w-'a'] == nullptr) {
                    node->children[w-'a'] = new TrieNode();
                }
                node = node->children[w-'a'];
            }
            node->s = word;
        }

        for (int r=0; r<row; r++) {
            for (int c=0; c<col; c++) {
                if (root->children[board[r][c]-'a'] == nullptr) continue;
                dfs(root->children[board[r][c]-'a'], r, c, board);
            }
        }

        return res;
    }
    void dfs(TrieNode* node, int r, int c, vector<vector<char>>& board) {
        if (!((node->s).empty())) {
            res.push_back(node->s);
            node->s = "";     // note the trick here
        }
        char tmp = board[r][c];
        board[r][c] = '#';
        for (auto &[dr, dc]:directions) {
            int nr = r+dr, nc = c+dc;
            if (0<=nr && nr<row && 0<=nc && nc<col && board[nr][nc] != '#' && node->children[board[nr][nc]-'a'] != nullptr) {
                dfs(node->children[board[nr][nc]-'a'], nr, nc, board);
            }
        }
        board[r][c] = tmp;
    }
};